### DISCUSSION TITLE
Exclusion list for a chemical reaction

### DISCUSSION THREAD

**Discussion by directysj** [mlund/faunus#378] [Q&A]
Hello @mlund ,

According to the manual of `Faunus`, currently, the `exclusion list/excluded pairs` are allowed for the `non-Bonded/bonded` interactions. Currently, a `chemical reaction` can happen by picking a atom randomly, as invoked in the` reaction list`. In a special case, we need to allow a `chemical reaction` at few specific sites of a residue (only) along the polymer chain. Thus, we need to create a list of residue positions along the polymer chain. As far as I know that it can not be done, correct me if I am wrong. Otherwise, can I ask your opinion/help?  

Thanking you in advance!!

**Comment by mlund** [comment]
Could a possible solution be to rename the particles in `atomlist` for the specific sites and simply use those names when defining the reaction? If so this could be done without any code changes. Alternatively, one could implement an exclusion list of index in the `rcmc` move that, if picked, would simply be ignored.

**Reply by directysj** [reply]
Thank you for the suggestions. The first one I would have tried but the problem is that the simulation is restarting from previously saved state file where there are no reactions on those sites. In this situation, I am afraid if renaming the particles can work! The second option can be tried, will it be similar to the exclusion list of non-bonding interaction pairs?

**Reply by directysj** [reply]
> Thank you for the suggestions. The first one I would have tried but the problem is that the simulation is restarting from previously saved state file where there are no reactions on those sites. In this situation, I am afraid if renaming the particles can work! The second option can be tried, will it be similar to the exclusion list of non-bonding interaction pairs?

One way may be that I need to change the names of those atoms in the state files as well.

**Reply by mlund** [reply]
For adding an exclusion list, we could target `SpeciationMove::bias()` and return infinity if one of the excluded atoms is part of the `Change` object. I'm thinking something like this:
``` yaml
moves:
  - rcmc:
    excluded:
      - protein: [0, 12, 19] # excluded sites in all "protein" molecules
```
The function `SpeciationMove::bias()` (speciation.cpp) should then investigate if any atoms in `Change.groups` is part of the exclusion list and then return inifinity if found, whereby the move will be rejected. I would recommend storing the exclusions as a `std::map<int, vector<int>>` where the key is the molecule id (`Group.id`) and `vector<int>` is the list of atoms to be excluded.

Todo:
1. Add private member `std::map<int, vector<int>> exclusions` as a private member in `Speciation`
2. Add `Speciation::readExclusions(const json &j)`
3. Add `bool Speciation::isExcluded(const Change &change)`
4. Add call to `Speciation::isExcluded` in `Speciation::bias` and return `pc::infty` if `true`
